lib/experimental/class-wp-html-walker-test.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+use PHPUnit\Framework\TestCase;
+
+require_once './class-wp-html-walker.php';
+
+final class WP_HTML_Walker_Tests extends TestCase {
+	const HTML_SIMPLE = '<div id="first"><span id="second">Text</span></div>';
+	const HTML_WITH_CLASSES = '<div class="main with-border" id="first"><span class="not-main bold with-border" id="second">Text</span></div>';
+
+	public function test_to_string_with_no_updates_returns_the_original_html() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$this->assertSame( self::HTML_SIMPLE, $w . '' );
+	}
+
+	public function test_finding_existing_tag_returns_the_walker_object() {
+		$w      = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$result = $w->find_next_tag();
+		$this->assertSame( $w, $result, 'Finding an existing tag returns the Walker object' );
+	}
+
+	public function test_finding_non_existing_tag_returns_false() {
+		$w      = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$result = $w->find_next_tag( 'p' );
+		$this->assertFalse( $result );
+	}
+
+	public function test_set_new_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->set_attribute( 'test-attribute', 'test-value' );
+		$this->assertSame( '<div test-attribute="test-value" id="first"><span id="second">Text</span></div>', $w . '' );
+	}
+
+	public function test_set_existing_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->set_attribute( 'id', 'new-id' );
+		$this->assertSame( '<div id="new-id"><span id="second">Text</span></div>', $w . '' );
+	}
+
+	public function test_remove_existing_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->remove_attribute( 'id' );
+		$this->assertSame( '<div ><span id="second">Text</span></div>', $w . '' );
+	}
+
+	public function test_remove_non_existing_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->remove_attribute( 'no-such-attribute' );
+		$this->assertSame( '<div id="first"><span id="second">Text</span></div>', $w . '' );
+	}
+
+	public function test_add_class_when_there_is_no_class_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->add_class( 'foo-class' );
+		$this->assertSame( '<div class="foo-class" id="first"><span id="second">Text</span></div>', $w . '' );
+	}
+
+	public function test_add_two_classes_when_there_is_no_class_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->add_class( 'foo-class' )->add_class( 'bar-class' );
+		$this->assertSame( '<div class="foo-class bar-class" id="first"><span id="second">Text</span></div>', $w . '' );
+	}
+
+	public function test_remove_class_when_there_is_no_class_attribute() {
+		$w = new WP_HTML_Walker( self::HTML_SIMPLE );
+		$w->find_next_tag()->remove_class( 'foo-class' );
+		$this->assertSame( '<div id="first"><span id="second">Text</span></div>', $w . '' );
```

Comment:
Similar as above, we can signal there are no changes applied with:
```suggestion
		$this->assertSame( self::HTML_SIMPLE, (string) $w );
```

Reply:
Good call! I just updated this.